test/extensions/io_socket/buffered_io_socket/buffered_io_socket_handle_impl_test.cc

```diff
@@ -0,0 +1,821 @@
+  EXPECT_CALL(*scheduable_cb_, scheduleCallbackNextIteration());
+  bool should_close = false;
+  io_handle_->initializeFileEvent(
+      dispatcher_,
+      [this, &should_close, handle = io_handle_.get(), &accumulator](uint32_t events) {
+        if (events & Event::FileReadyType::Read) {
+          auto result = io_handle_->recv(buf_.data(), buf_.size(), 0);
+          if (result.ok()) {
+            accumulator += absl::string_view(buf_.data(), result.rc_);
+          } else if (result.err_->getErrorCode() == Api::IoError::IoErrorCode::Again) {
+            ENVOY_LOG_MISC(debug, "read returns EAGAIN");
+          } else {
+            ENVOY_LOG_MISC(debug, "will close");
+            should_close = true;
+          }
+        }
+      },
+      Event::FileTriggerType::Edge, Event::FileReadyType::Read);
+  scheduable_cb_->invokeCallback();
+
+  // Not closed yet.
+  ASSERT_FALSE(should_close);
+
+  EXPECT_CALL(*scheduable_cb_, scheduleCallbackNextIteration());
+  io_handle_peer_->shutdown(ENVOY_SHUT_WR);
+
+  ASSERT_TRUE(scheduable_cb_->enabled());
+  scheduable_cb_->invokeCallback();
+  ASSERT_FALSE(should_close);
+  EXPECT_EQ(4, accumulator.size());
+  io_handle_->close();
+  io_handle_->resetFileEvents();
+}
+
+TEST_F(BufferedIoSocketHandleTest, TestRepeatedShutdownWR) {
+  EXPECT_EQ(io_handle_peer_->shutdown(ENVOY_SHUT_WR).rc_, 0);
+  EXPECT_EQ(io_handle_peer_->shutdown(ENVOY_SHUT_WR).rc_, 0);
+}
+
+TEST_F(BufferedIoSocketHandleTest, TestShutDownOptionsNotSupported) {
+  ASSERT_DEBUG_DEATH(io_handle_peer_->shutdown(ENVOY_SHUT_RD), "");
+  ASSERT_DEBUG_DEATH(io_handle_peer_->shutdown(ENVOY_SHUT_RDWR), "");
+}
+
+TEST_F(BufferedIoSocketHandleTest, TestWriteByMove) {
+  Buffer::OwnedImpl buf("0123456789");
+  auto result = io_handle_peer_->write(buf);
+  EXPECT_TRUE(result.ok());
+  EXPECT_EQ(10, result.rc_);
+  auto& internal_buffer = getWatermarkBufferHelper(*io_handle_);
+  EXPECT_EQ("0123456789", internal_buffer.toString());
+  EXPECT_EQ(0, buf.length());
+}
+
+// Test write return error code. Ignoring the side effect of event scheduling.
+TEST_F(BufferedIoSocketHandleTest, TestWriteErrorCode) {
+  Buffer::OwnedImpl buf("0123456789");
+  Api::IoCallUint64Result result{0, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+
+  {
+    // Populate write destination with massive data so as to not writable.
+    auto& internal_buffer = getWatermarkBufferHelper(*io_handle_peer_);
+    internal_buffer.setWatermarks(1024);
+    internal_buffer.add(std::string(2048, ' '));
+
+    result = io_handle_->write(buf);
+    ASSERT_EQ(result.err_->getErrorCode(), Api::IoError::IoErrorCode::Again);
+    EXPECT_EQ(10, buf.length());
+  }
+
+  {
+    // Write after shutdown.
+    io_handle_->shutdown(ENVOY_SHUT_WR);
+    result = io_handle_->write(buf);
+    ASSERT_EQ(result.err_->getErrorCode(), Api::IoError::IoErrorCode::UnknownError);
+    EXPECT_EQ(10, buf.length());
+  }
+
+  {
+    io_handle_peer_->close();
```

Comment:
The "io_handle_->shutdown(ENVOY_SHUT_WR);" on line 483 above interacts badly with this part of the test.

io_handle_->write will fail even if you don't execute io_handle_peer_->close() because of the io_handle_->shutdown call.